core/controllers/admin_test.py

```diff
@@ -963,6 +963,7 @@ def test_update_param_rules_with_param_name_of_non_string_type_returns_400(
             expected_status_int=400
         )
         error_msg = (
+            'At \'http://localhost/adminhandler\' these errors are happening:\n'
```

Comment:
I'm not sure a helper function would reduce code duplication all that much. We'd be going from

```python
error_msg = (
    'At \'http://localhost/adminhandler\' these errors are happening:\n'
    'Schema validation for \'platform_param_name\' failed: Expected '
    'string, received 123')
```

to

```python
error_msg = getController_error_msg_prefix(
    'http://localhost/adminhandler',
    'Schema validation for \'platform_param_name\' failed: Expected '
    'string, received 123')
```

(I think if we did go with the helper function, passing the error message into the helper would be better so that the helper function can handle the string concatenation.)

To my eye the latter isn't really any cleaner. I followed the first approach because that's what we do elsewhere in the codebase. For example, here's a base controller test in `develop`:

https://github.com/oppia/oppia/blob/1f95c06707d783aa040de0676af1e79aaa841c80/core/controllers/base_test.py#L1866-L1871

Unless we want to change all existing patterns like this, I think we should stay consistent.

One option would be to use `assertIn` instead of `assertEquals` and not check the URL. I did this in some places where the URL is unpredictable (e.g. contains a randomly-generated ID). Wherever possible though, I think it's important to assert that the error messages include the URL because the absence of a URL made errors harder to debug. That's the primary reason for changing the base controller to add URLs, so those URLs should be asserted in the tests.

Reply:
Re @StephenYu2018's note, I'm OK with the hardcoding. I think we should keep the tests as simple to read as possible, so in general I'm fine with hardcoding "expected" things.

Reply:
Will these tests fail if any of the referenced URLs were to change?

Reply:
Yes, they would. To avoid that, we could construct the URLs from constants in [`feconf.py`](https://github.com/oppia/oppia/tree/develop/core/feconf.py) (though I haven't checked whether all the constants we'd need are present there). To @seanlip's point however, this change would make the tests a little harder to read.

I searched through the existing tests in develop, and it looks like most tests that assert URLs do so by writing out the URL in the string literal. Since that's the current norm in the codebase, if we decide to deviate from that norm here, I think we should change all the other cases too. That would be a much broader change than needed to address #19858, so maybe that should be a follow-up PR? It'll also be easier to review this refactor in a follow-up PR where there are no functional changes.

Reply:
@U8NWXD That's fair. I think we should put off such a potential implementation for a future PR.